### PR TITLE
Logging: Add customizable logging backends and fmtlib based macros

### PR DESCRIPTION
**Pull request by jroweboy**
Supersedes https://github.com/citra-emu/citra/pull/2707

TL;DR:
* Change the logging backend to support multiple sinks through the Backend Interface
* Add a new set of logging macros to use fmtlib instead.
* Qt: Compile as GUI application on windows to make the console hidden by
default. Add filter configuration and a button to open log location.
* SDL: Migrate to the new logging macros

I tried to apply the lessons learned from the last logging PR. This time I built my own sinks instead of trying to attach spdlog sinks. The new logger splits sinks out from sources by using the MPSC queue and a worker thread. The worker thread will check for new logs and call `Write` for each `Backend` that is registered.

How logging works now:
* Setup for each frontend: 
  * Load the settings, and configures the global logging filter (default "*:Info").
  * Get the log location from `GetUserPath(D_LOGS_IDX)` and creates the folder if missing.
  * Add Backends to the list of backends. Both frontends do a color console and a file backend
* printf or fmt lib macro is called. 
  * builds a `Entry` which gets added to the MPSC queue (so its thread safe to call the logging macros)
* worker thread:
  * Consumes from the queue. Get the `Entry` and passes it to each configured backend
  * Backends can use the convenience functions in `text_formatting` to build the correct format string.

After this is merged, we will need to do what was planned for the original PR and migrate the old logging macros to the new one, followed by a final commit to rename it back to the original name.

Sorry for random format changes. I'm still using clang-format 6 while waiting for https://github.com/citra-emu/citra/pull/3410 to get merged ;) hint hint

Debug tab changed to add logging stuff
citra-qt_2018-02-19_18-30-24

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/3449)

### CODE REVIEW

File: src/common/logging/backend.cpp [hunk 1]
```diff
@@ -4,16 +4,108 @@
...
+    }
+
+private:
+    Impl() : running(true) {
+        backend_thread = std::async(std::launch::async, [&] {
+            using namespace std::chrono_literals;
+            Entry entry;
+            while (running) {
+                if (!message_queue.Pop(entry)) {
+                    std::this_thread::sleep_for(1ms);
```
Comment (jroweboy) [inline src/common/logging/backend.cpp]: Is 1ms too much to wait for new logs?

File: src/citra_qt/debugger/console.cpp [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+#include <windows.h>
+
+#include <wincon.h>
+#endif
+
+#include "citra_qt/debugger/console.h"
+#include "citra_qt/ui_settings.h"
+
+namespace Debugger {
+void ToggleConsole() {
```
Comment (j-selby) [inline src/citra_qt/debugger/console.cpp]: Is there a reason why this couldn't just be a a Qt window? Seems a bit weird to be adding win32 conhost bindings in the Qt frontend

Comment (jroweboy) [reply src/citra_qt/debugger/console.cpp]: Yup. I wanted to make a QT window, but that'll be for another day. I spent about 7 1/2 hours on this and I imagine a good UI for it would take just as long :)

File: src/common/logging/backend.cpp [hunk 3]
```diff
@@ -113,45 +205,65 @@ const char* GetLevelName(Level log_level) {
...
 
-void LogMessage(Class log_class, Level log_level, const char* filename, unsigned int line_nr,
+void LogMessage(Class log_class, Level log_level, const char* filename, unsigned int line_num,
                 const char* function, const char* format, ...) {
-    if (filter != nullptr && !filter->CheckMessage(log_class, log_level))
+    auto filter = Impl::Instance().GetGlobalFilter();
+    if (!filter.CheckMessage(log_class, log_level))
         return;
-
+    std::array<char, 4 * 1024> formatting_buffer;
```
Comment (lioncash) [inline src/common/logging/backend.cpp]: This seems like something that should be managed by the logger itself, so it's not continually placed on the stack every time a message is logged, but it was already like this, so feel free to ignore it.

File: src/common/logging/backend.cpp [hunk 4]
```diff
@@ -113,45 +205,65 @@ const char* GetLevelName(Level log_level) {
...
 
     static steady_clock::time_point time_origin = steady_clock::now();
 
-    std::array<char, 4 * 1024> formatting_buffer;
-
     Entry entry;
     entry.timestamp = duration_cast<std::chrono::microseconds>(steady_clock::now() - time_origin);
     entry.log_class = log_class;
     entry.log_level = log_level;
+    entry.filename = std::string(Common::TrimSourcePath(filename));
```
Comment (lioncash) [inline src/common/logging/backend.cpp]: This can just be assigned to `filename` without the std::string constructor call.

File: src/common/logging/backend.cpp [hunk 5]
```diff
@@ -113,45 +205,65 @@ const char* GetLevelName(Level log_level) {
...
+    entry.line_num = line_nr;
+    entry.function = std::string(function);
```
Comment (lioncash) [inline src/common/logging/backend.cpp]: Ditto

File: src/common/logging/backend.cpp [hunk 6]
```diff
@@ -4,16 +4,108 @@
...
+        auto it = std::find_if(backends.begin(), backends.end(), [&backend_name](const auto& i) {
+            return i->GetName() == backend_name;
+        });
+        if (it == backends.end())
+            return nullptr;
+        return it->get();
+    }
+
+private:
+    Impl() : running(true) {
```
Comment (lioncash) [inline src/common/logging/backend.cpp]: This can be initialized in the class:

```cpp
class Thing
{
private:
    std::atomic_bool running{true};
};

File: src/common/logging/backend.cpp [hunk 7]
```diff
@@ -4,16 +4,108 @@
...
+                    continue;
+                }
+                for (const auto& backend : backends) {
+                    backend->Write(entry);
+                }
+            }
+        });
+    }
+    ~Impl() {
+        if (running) {
```
Comment (lioncash) [inline src/common/logging/backend.cpp]: Does this if statement matter? If the object it being torn down, I doubt the boolean value matters much.

File: src/common/logging/backend.cpp [hunk 8]
```diff
@@ -113,45 +205,65 @@ const char* GetLevelName(Level log_level) {
 }
 
 Entry CreateEntry(Class log_class, Level log_level, const char* filename, unsigned int line_nr,
-                  const char* function, const char* format, va_list args) {
-    using std::chrono::steady_clock;
+                  const char* function, const std::string& message) {
```
Comment (lioncash) [inline src/common/logging/backend.cpp]: I would take `message` by value here and std::move it into `entry.message`. This can avoid copies, as it can be moved into from the call site. e.g. The use in `FmtLogMessage` can utilize std::string's move constructor as-is, if this is passed by value instead of const reference.

File: src/common/logging/text_formatter.cpp [hunk 9]
```diff
@@ -18,50 +18,26 @@
...
+                       class_name, level_name, entry.filename, entry.function, entry.line_num,
+                       entry.message);
 }
 
 void PrintMessage(const Entry& entry) {
-    std::array<char, 4 * 1024> format_buffer;
-    FormatLogMessage(entry, format_buffer.data(), format_buffer.size());
-    fputs(format_buffer.data(), stderr);
-    fputc('\n', stderr);
+    auto str = FormatLogMessage(entry) + "\n";
```
Comment (lioncash) [inline src/common/logging/text_formatter.cpp]: \n can be a character literal, which can be cheaper to append than a string: `'\n'`

File: src/common/logging/backend.cpp [hunk 10]
```diff
@@ -4,16 +4,108 @@
...
+
+    const Filter& GetGlobalFilter() const {
+        return filter;
+    }
+
+    void SetGlobalFilter(const Filter& f) {
+        filter = f;
+    }
+
+    Backend* GetBackend(const std::string& backend_name) {
```
Comment (lioncash) [inline src/common/logging/backend.cpp]: This should probably return a boost::optional instead of a raw pointer, so nullptr isn't used as a synonym for "not found".

File: src/common/logging/backend.cpp [hunk 11]
```diff
@@ -4,16 +4,108 @@
...
+public:
+    static Impl& Instance() {
+        static Impl backend;
+        return backend;
+    }
+
+    Impl(Impl const&) = delete;
+    const Impl& operator=(Impl const&) = delete;
+
+    Common::MPSCQueue<Log::Entry>& GetQueue() {
```
Comment (lioncash) [inline src/common/logging/backend.cpp]: Does it make sense to expose the queue itself, instead of providing a `PushEntry` function that forwards the entry to the queue as part of the API?

File: src/common/logging/backend.cpp (hunk 10 above)
Comment (jroweboy) [reply src/common/logging/backend.cpp]: I was originally avoiding linking boost to common, but then i just learned that it was already linked to common.

File: src/common/logging/backend.cpp [hunk 12]
```diff
@@ -113,45 +203,65 @@ const char* GetLevelName(Level log_level) {
...
-    std::array<char, 4 * 1024> formatting_buffer;
-
     Entry entry;
     entry.timestamp = duration_cast<std::chrono::microseconds>(steady_clock::now() - time_origin);
     entry.log_class = log_class;
     entry.log_level = log_level;
+    entry.filename = Common::TrimSourcePath(filename);
+    entry.line_num = line_nr;
+    entry.function = function;
+    entry.message = message;
```
Comment (lioncash) [inline src/common/logging/backend.cpp]: message can be std::moved here now

File: src/common/logging/backend.cpp (hunk 7 above)
Comment (MerryMage) [reply src/common/logging/backend.cpp]: `~future()` (of `backend_thread`) would block since the work initiated by `std::async` hasn't terminated, as far as I understand it.

Refs:
* [The release of the shared state may block](http://eel.is/c++draft/futures.state#5.3).
* [`~future` releases shared state](http://eel.is/c++draft/futures.unique_future#9.1).

**Comment by jroweboy** [comment]
Summary of the new updates. Thanks to @chris062689 for suggesting some of these improvements.

* Uses PopWait to reduce the amount of busy waiting if there aren't many new logs
* Opens the log file as shared on windows, letting other programs read the logs, but not write to them while citra is running
* Flushes the logs to disk if a log >= error arrives

**Comment by chris062689** [comment]
I'll test these changes tonight, thanks @jroweboy for those tweaks, I think it improves usability.

File: src/citra_qt/configuration/configure_debug.ui [hunk 13]
```diff
@@ -72,6 +72,47 @@
...
+         <widget class="QLineEdit" name="log_filter_edit"/>
+        </item>
+       </layout>
+      </item>
+      <item>
+       <layout class="QHBoxLayout" name="horizontalLayout_2">
+        <item>
+         <widget class="QCheckBox" name="toggle_console">
+          <property name="text">
+           <string>Show Log Console (Windows Only)</string>
```
[the remark this answers is not in the document]
Comment (jroweboy) [reply src/citra_qt/configuration/configure_debug.ui]: If you read the comment in the source code, the idea is to replace this feature with a QT console window entirely. I don't see a point in dealing with platform specific hiding and showing a single checkbox on the debug tab.

**Comment by jroweboy** [comment]
>this could easily lead to TB big log files if you play for a few hours. Ofc this could be some sort of hidden child protection, but Imo it's better to have the option to disable logging to a file.

If this is really a problem, then we should move some of the Debug logs that we have into Trace. 

There should not be a way to disable the log entirely imo. This will lead to scenarios where we can't debug issues because people downloaded a build with logging off and they are unable to figure out where to turn it back on. There is already a log level filter, so someone could just easily set it to Critical now, but turning off logs is a non-goal entirely to me.

**Comment by wwylele** [comment]
> > this could easily lead to TB big log files if you play for a few hours. Ofc this could be some sort of hidden child protection, but Imo it's better to have the option to disable logging to a file.
> 
> If this is really a problem, then we should move some of the Debug logs that we have into Trace.
> 

Here example that you couldn't easily fix with Trace level:

When a game jumps to pc = 0 for some reason (Panic break, wrong branching, null function pointer calling...), citra will start to log `Unmapped Read32` (Error level) for _every_ single CPU instructions, which is very fast (especially in JIT mode because dynarmic would try to go through the whole code first). Now with log window hidden by default, user would only notice that the game freezes, and citra window unable to close (because the emu thread is locked in the CPU loop), and won't have an idea that the log file is being filled up very quickly.

File: src/common/threadsafe_queue.h [hunk 14]
```diff
@@ -75,6 +83,25 @@ class SPSCQueue {
...
     }
 
+    /**
+     * Waits up to timeout for data to be Pushed to the queue. Push uses a condition variable to
+     * signal the waiting thread, but only if NeedSize = true. Returns false if the timeout is
+     * triggered. If the condition variable is signalled, returns the value from Pop
+     * @param T In parameter to store the value if this method returns true
+     * @param timeout Time in milliseconds to wait for a signal from a Push
+     */
+    bool PopWait(T& t, u64 timeout = 500) {
```
[the remark this answers is not in the document]
Comment (jroweboy) [reply src/common/threadsafe_queue.h]: Since this one is in common, I wanted to avoid pulling in a dependency on chrono. In actuality, its only used in two places so its not a big deal if you want me to add it.

File: src/common/logging/backend.h [hunk 15]
```diff
@@ -22,15 +26,81 @@ struct Entry {
...
+public:
+    virtual ~Backend() = default;
+    virtual void SetFilter(const Filter& new_filter) {
+        filter = new_filter;
+    }
+    virtual const char* GetName() const = 0;
+    virtual void Write(const Entry& entry) = 0;
+
+private:
+    Filter filter;
```
[the remark this answers is not in the document]
Comment (jroweboy) [reply src/common/logging/backend.h]: I like that idea better. Will change that.

**Comment by jroweboy** [comment]
> This kind of log exists for many games (most of them is due to wrong region configuration though)

Fair point. Any suggestions on how best to handle the large log file then?

**Comment by BreadFish64** [comment]
`This will lead to scenarios where we can't debug issues because people downloaded a build with logging off and they are unable to figure out where to turn it back on.`
We don't support unofficial builds, so I don't really see how that's our problem.

**Comment by B3n30** [comment]
IMO the best solution is to have an option to enable the logging to file by default and have a option in config/debug to enable logging to file. If someone ask for support/has an issue we can tell them to enable logging to file, run the game till the issue occurs and then send us the log.

Another solution would be to generate a warning if the log file reaches a certain size. But no idea on how to continue at that point.

**Comment by jroweboy** [comment]
Does any one have any issue with cutting off the file log at 50MB? if your filter is setup such that you spam more than 50MB of data then you should likely modify the filter.

I still disagree that we should allow disabling file logging. Logging is meant to be a post mortem debugging tool, so we should have it on by default to catch errors when they happen, and not turn it on and attempt to replicate it.

**Comment by daniellimws** [comment]
How about compressing the current log file at say 50MB, then create a new one (like part 2), so as to have "multi-parts" of the log? (Not sure if this would introduce dependencies)

**Comment by adityaruplaha** [comment]
Well, for many commercial games, the log spam is very high. So it might reach well over 50 MB.

@daniellimws I think that's a great idea.

Well, it's just a suggestion, but can't we just check if the Read32 exception (or similar) has already been thrown? Then we can just print a `"..."`. Then it'll handle the problem quite well. But it'll be a little like a hack.

**Comment by jroweboy** [comment]
I don't have the time to put in enough effort to satisfy everyone anymore on this PR. Heres the patch that I was writing, its not complete, but it should be pretty good. If someone wants to pick this up and finish it go ahead.

[diff.txt](https://github.com/citra-emu/citra/files/1764069/diff.txt)

In the diff, I've added a condition var to backend impl instead of in the queue so that it can wait on both variables instead of spin waiting on running. I also limited the amount written to disk in the file backend to 50 MB. Addressed most of wwylele's comments. After shutting down, it tries to write at least 100 more logs to clear out the queue before finally closing. Was working on removing and adding the ConsoleBackend when showing/hiding the console as a test of that system and it didn't work. Don't care to fix that but its something that the next person could look at or eventually if I get time again I'll look at it. Got more important PRs to work on in the meantime.

@daniellimws I recall you wanted something to pick up. I'd hate to pass the buck off to someone, but you are more than willing to resubmit this PR with those issues addressed if you want.

**Comment by jroweboy** [comment]
FYI I'm closing this because it conflicts with multiplayer UI and since its not getting merged soon, i might as well just close the UI gets fixed and merged
